src/kernel/MomentumOpenAdvDiffElemKernel.C

```diff
@@ -0,0 +1,370 @@
+
+  elemDataPreReqs.add_coordinates_field(*coordinates_, BcAlgTraits::nDim_, CURRENT_COORDINATES);
+  elemDataPreReqs.add_gathered_nodal_field(*velocityNp1_, BcAlgTraits::nDim_);
+  elemDataPreReqs.add_gathered_nodal_field(*velocityRTM_, BcAlgTraits::nDim_);
+  elemDataPreReqs.add_gathered_nodal_field(*viscosity_, 1);
+  elemDataPreReqs.add_gathered_nodal_field(*density_, 1);
+ 
+  if ( shiftedGradOp_ )
+    elemDataPreReqs.add_master_element_call(SCS_SHIFTED_FACE_GRAD_OP, CURRENT_COORDINATES);
+  else
+    elemDataPreReqs.add_master_element_call(SCS_FACE_GRAD_OP, CURRENT_COORDINATES);
+
+  // never shift properties
+  get_face_shape_fn_data<BcAlgTraits>([&](double* ptr){meFC->shape_fcn(ptr);}, vf_shape_function_);
+  get_scs_shape_fn_data<BcAlgTraits>([&](double* ptr){meSCS->shape_fcn(ptr);}, v_shape_function_);
+}
+
+template<typename BcAlgTraits>
+MomentumOpenAdvDiffElemKernel<BcAlgTraits>::~MomentumOpenAdvDiffElemKernel()
+{}
+
+template<typename BcAlgTraits>
+void
+MomentumOpenAdvDiffElemKernel<BcAlgTraits>::execute(
+  SharedMemView<DoubleType**> &lhs,
+  SharedMemView<DoubleType *> &rhs,
+  ScratchViews<DoubleType> &faceScratchViews,
+  ScratchViews<DoubleType> &elemScratchViews)
+{
+  DoubleType w_uBip[BcAlgTraits::nDim_];
+  DoubleType w_uScs[BcAlgTraits::nDim_];
+  DoubleType w_uBipExtrap[BcAlgTraits::nDim_];
+  DoubleType w_uspecBip[BcAlgTraits::nDim_];
+  DoubleType w_coordBip[BcAlgTraits::nDim_];
+  DoubleType w_nx[BcAlgTraits::nDim_];
+
+  // face
+  SharedMemView<DoubleType*>& vf_viscosity = faceScratchViews.get_scratch_view_1D(*viscosity_);
+  SharedMemView<DoubleType**>& vf_velocityNp1 = faceScratchViews.get_scratch_view_2D(*velocityNp1_);
+  SharedMemView<DoubleType**>& vf_bcVelocity = faceScratchViews.get_scratch_view_2D(*velocityBc_);
+  SharedMemView<DoubleType***>& vf_Gjui = faceScratchViews.get_scratch_view_3D(*Gjui_);
+  SharedMemView<DoubleType**>& vf_coordinates = faceScratchViews.get_scratch_view_2D(*coordinates_);
+  SharedMemView<DoubleType**>& vf_exposedAreaVec = faceScratchViews.get_scratch_view_2D(*exposedAreaVec_);
+  SharedMemView<DoubleType*>& vf_openMassFlowRate = faceScratchViews.get_scratch_view_1D(*openMassFlowRate_);
+  
+  // element
+  SharedMemView<DoubleType**>& v_coordinates = elemScratchViews.get_scratch_view_2D(*coordinates_);
+  SharedMemView<DoubleType**>& v_velocityNp1 = elemScratchViews.get_scratch_view_2D(*velocityNp1_);
+  SharedMemView<DoubleType**>& v_vrtm = elemScratchViews.get_scratch_view_2D(*velocityRTM_);
+  SharedMemView<DoubleType**>& v_viscosity = elemScratchViews.get_scratch_view_1D(*viscosity_);
+  SharedMemView<DoubleType**>& v_density = elemScratchViews.get_scratch_view_1D(*density_);
+  SharedMemView<DoubleType***>& v_dndx = shiftedGradOp_
+    ? elemScratchViews.get_me_views(CURRENT_COORDINATES).dndx_shifted_fc_scs
+    : elemScratchViews.get_me_views(CURRENT_COORDINATES).dndx_fc_scs;
+
+  // fixme: need master elements
+  MasterElement* meFC = NULL;
```

Comment:
I would like to avoid something like:

`MasterElement *meSCS = sierra::nalu::MasterElementRepo::get_surface_master_element(BcAlgTraits::elemTopo_);`

Reply:
Why?  
Save it off during the constructor instead?

Reply:
Yes that's what I was going to suggest. The master-elements are set/available in the kernel constructor, they could be set as members at that point.

Reply:
Right... Done:)